### ISSUE TITLE
Allow default value for union fields

### ISSUE THREAD

**Issue by Exonorid** [ziglang/zig#9524]
Currently, trying to compile something like
```zig
pub fn main() void {
    const Foo = struct{a: ?u8, b: ?u8, c: ?u8, d: ?u8};

    var test_foo = Foo{.a = 5, .b = 6, .c = null, .d = 9};
    inline for(std.meta.fields(Foo)) |field_info| {
        if(@field(test_foo, field_info.name) == null) continue;
        doStuff(@field(test_foo, field_info.name));
    }
}
```
won't work because the `continue` is based on a runtime value. However, there's no reason not to allow this behavior at runtime by unwrapping it into something like
```c
if(test_foo.a == null) goto for_b;
doStuff(test_foo.a);
for_b:
if(test_foo.b == null) goto for_c;
doStuff(test_foo.b);
//etc
```

Although I will admit that this example is contrived, I just wanted a simple example to illustrate my point. For more complicated control flow, it can greatly simplify code and make it more readable.

**Comment by Exonorid** [comment]
Sorry about the title, I wasn't paying attention

**Comment by marler8997** [comment]
I think this breaks Zig.   In your case, `doStuff` looks like it might be a runtime function call, but if it's a comptime function call, then whether or not it gets called depends on the result of the `if/continue`, which it can't determine at comptime.  I suppose if everything following the continue was runtime only, maybe it could work?

**Comment by codehz** [comment]
> which it can't determine at comptime.
yes, but I think it is possible to jump(or "goto") to the end of the whole inline for statement, and that's can be determined at comptime.

**Comment by zzyxyzz** [comment]
> In your case, doStuff looks like it might be a runtime function call, but if it's a comptime function call, then whether or not it gets called depends on the result of the if/continue, which it can't determine at comptime.

That looks reasonably well-defined to me. If `doStuff` is runtime, it is left as is. If it's comptime, and the condition guarding it cannot be evaluated, then `doStuff` is always run. It is only elided if it can be shown to be definitely unreachable. That's how it works in other cases as well:

```zig
const a = false;
if (a) @compileError("!"); // elided

var b = false;
if (b) @compileError("!"); // throws
```

That said, I lean towards keeping the "unroll fully or error" semantics of inline loops, as opposed to "unroll what you can". Otherwise, nothing would prevent you from accidentally generating a bunch of unwanted runtime checks, because you forgot to mark some `var` as comptime, or call a function that wasn't quite as comptime as expected.

**Comment by AssortedFantasy** [comment]
> I think this breaks Zig.

Not really, the semantics of this seem pretty clear cut. You can currently work around what is probably a bug like this:

```zig
inline for (something) |stuff| {
  // expressions
  if (!condition) continue;
  // more expressions
}
// exactly equivalent to <=>
inline for (something) |stuff| {
  // expressions
  if (condition) {
    // more expressions
  }
}
```

```zig
inline for (something) |stuff| {
  // expressions
  if (!condition) break;
  // more expressions
}
// exactly equivalent to <=>
var is_breaked = false;
inline for (something) |stuff| {
  if (!is_breaked) {
    // expressions
    if (condition) {
      // more expressions
    } else {
      is_breaked = true;
    }
  }
}
```
Not sure about the coden on the simulated "break". It'd be more effient with a single jump to the end rather than
what naively codegens into a chain of compare and jumps against the same condition.

**Comment by CurtisFenner** [comment]
As a workaround that doesn't force you to rearrange how your `if`s are structured, you can use two "runtime" blocks/loops sandwiching the `inline` loop to get `break`/`continue`

```
outer: {
    inline for (arr) |e| {
        inner: {
                ......
                outer: break; // equivalent to a "runtime break"
                ......
                inner: break; // equivalent to a "runtime continue"
                ......
        }
    }
}
```

however you'll probably run into #2727 which is apparently still open (I have not verified this recently).

Using conditional `break` and `continue` doesn't affect unrolling at all; a `break` is just an unconditional jump to the outer block, and a `continue` is just an unconditional jump to the end of the current iteration.

Using explicit blocks is OK as a workaround, but it requires two pointless labeled blocks. Other conditional runtime logic, like a `if (...) return` with an `inline for`, (or labeled breaks!, as in the workaround) isn't blocked, yet the effect is exactly the same. I don't think that blocking an unmarked `break` and `continue` when it happens to refer to an `inline` loop buys much at all; if there is a need to be sure that they're not becoming unconditional jumps, you can have an `inline break`/`comptime break` construct that throws an error, instead.

Given that it's straightforward but annoying to apply this structure, I think the language should permit this construct with the intuitive syntax -- simply `break` and `continue` as normal.

**Comment by zzyxyzz** [comment]
#2727, #1609 and a host of other issues are still open, so you cannot `break` out of inline loops reliably. stage1 just doesn't like runtime control flow in inline loops *at all*. However, for stage2, we should seriously think about what the desired semantics is. On the one hand, a break or continue that is incorrectly deferred to runtime is merely a performance issue. On the other, explicit unrolling for perfomance is one of the major uses of inline loops, so we shouldn't dismiss silent failures to optimize as unimportant.

There's also the question whether something like this should be allowed:
```zig
inline while (runtime_cond) {
    if (comptime_cond) break; // causes loop to terminate without exhausting branch quota
    // do something ...
}
```
After all, this is exactly equivalent to
```zig
inline while (true) {
    if (runtime_cond) break;
    if (comptime_cond) break;
    // do something ...
}
```
If the latter is permitted, why not the former?

Personally, I would prefer it if direct `break` and `continue` of an inline loop were only allowed within comptime-evaluated control paths. Breaking and continuing at runtime would require the use of labelled break as shown above. This way, we don't lose any functionality, but gain a visual distinction between runtime and comptime control flow, and also get error messages when something that's supposed to be comptime isn't.

In short:
```zig
outer: {
    inline while (cond1) inner: {
        if (cond2) break;
        if (cond3) continue;
        if (cond4) break :outer; // break equivalent
        if (cond5) break :inner; // continue equivalent
        if (cond6) return;
    )
}
// cond 1-3: comptime only
// cond 4-5: can be runtime
```

**Comment by AssortedFantasy** [comment]
@zzyxyzz I don't think your reasoning is correct.
```zig
inline while (true)
```
isn't a thing, wont ever ever compile since it requires an infinite branch quota. That's why the thing inside of the `inline while(cond)` has to be comptime known. 

Inlines while don't have very complicated semantics and I don't think we need to make such complex conditions on break and continues. They should work fine at runtime or comptime. See my comment for the semantics.

The reason is because "inline while" basically is just a copy paste of a block some compile time known number of times. Thats it. A runtime break is just a jump to the end of the blocks, a runtime continue is just a jump to the end of the current copy.

A comptime break lets you avoid emitting (statically eliminates) the rest of the blocks, a comptime continue lets you avoid emitting the rest of the current block.

**Comment by zzyxyzz** [comment]
@AssortedFantasy
I was just pointing out the equivalence between the loop condition and a conditional break in the loop body. So if we want the loop condition to be comptime-known, we probably intend breaks and continues to be comptime as well. And vice versa, if runtime breaks and continues are allowed in an otherwise comptime loop, then why not also in the loop condition, provided that the loop still terminates at comptime?

I'm also not arguing that the semantics of mixed runtime/comptime conditional breaks is somehow ill defined. It's just that what gets included in the final binary and what doesn't is very much part of the intent of an inline loop, and muddying this distinction is not a great thing to do, IMO.

**Comment by AssortedFantasy** [comment]
@zzyxyzz This is exactly what I disagree about. I don't think they are equivalent, hence we can have it both ways. 
I don't think just because runtime breaks work somehow implies the loop conditions are allowed to be runtime.

Specifically, I think this is poorly defined:
>inline while (runtime_cond) {
    if (comptime_cond) break; // causes loop to terminate without exhausting branch quota
    // do something ...
}

The reason I don't think this is well defined, is because why should this start unrolling in the first place?
The compiler uses inline whiles to unroll statements. If you don't know `runtime_cond` at compile time, the compiler can't choose to either start unroll, or not start unrolling. Its not `true` or `false`.

>inline while (runtime_cond) <=> inline while (true) // <-- this right here is what I disagree with, why do we assume true.

At compile time you need to decide if you unroll or not. If make it so that "runtime_conds" are assumed true then yeah, sure, everything becomes inconsistent, but if you don't have that assumption then nothing is muddied imo.

**Comment by zzyxyzz** [comment]
@AssortedFantasy,
> The reason I don't think this is well defined, is because why should this start unrolling in the first place? The compiler uses inline whiles to unroll statements. If you don't know runtime_cond at compile time, the compiler can't choose to either start unroll, or not start unrolling. Its not true or false.

The compiler starts unrolling because it sees the `inline` attribute. If the loop condition cannot be evaluated at comptime, a runtime conditional break is inserted at the start of every unrolled instance. Runtime-conditional `break`s and `continue`s within the body are handled in the same fashion.

The unrolling terminates either when the loop condition evaluates to false at comptime, or when a `break` is reached on a comptime control path within the body. If neither happens, the compiler eventually hits the branch quota and aborts with an error.

Thus, from a purely mechanical perspective, all combinations of runtime and comptime control flow in inline loops are well-defined and at least somewhat sensible. Whether they conform to *expectations* is another matter, though. Personally, I tend to think about loops in desugared terms, and therefore expect the loop condition and conditional breaks to be treated equally. Which does not make the expectations of others any less valid, of course.

**Comment by AssortedFantasy** [comment]
@zzyxyzz 
> If the loop condition cannot be evaluated at comptime, [we assume its true/ start unrolling anyway] and a runtime conditional break is inserted at the start of every unrolled instance. Runtime-conditional breaks and continues within the body are handled in the same fashion.

I suppose we will just have to agree to disagree then. This entire conversation has just been more or less just me repeating the same thing about how I'd say this isn't expected / isn't related to runtime breaks and continues / isn't very sensible even if it is _possible_ that we define `inline while` semantics like this.

If you'd like to talk more feel free to message me on the Zig discord as KingStnap. Otherwise we might as well not bloat this issue further with more of the same thing.
